File: Kattis/openkattis/fifty_shades_pink.hs
-- Kattis, Fifty Shades of Pink

import Data.Char (toLower)
import Data.List (isInfixOf)

test1 =
  [ "pink"
  , "tequilaSunrose"
  , "mExicanPInK"
  , "Coquelicot"
  , "turqrose"
  , "roSee"
  , "JETblack"
  , "pink"
  , "babypink"
  , "pInKpinkPinK"
  , "PInkrose"
  , "lazerlemon" ]

test2 =
  [ "roose"
  , "rosse"
  , "pinnk"
  , "screw"
  , "cerise"
  , "magenta" ]

haspink :: String -> Bool
haspink = isInfixOf "pink"

hasrose :: String -> Bool
hasrose = isInfixOf "rose"

isvalid :: String -> Bool
isvalid s = haspink s || hasrose s

lowerword :: String -> String
lowerword = fmap toLower

solve :: [String] -> Either Int String
solve cs
  | nums == 0 = Right "I must watch Star Wars with my daughter"
  | otherwise = Left nums
  where nums = length $ filter isvalid $ fmap lowerword cs
  
main :: IO ()
main = interact $ either show id . solve . tail . lines

File: Kattis/openkattis/fifty_shades_pink.ml
(* Kattis, fifty shades of pink *)

module Either = struct
  type ('a,'b) t =
    | Left of 'a
    | Right of 'b

  let either f g = function
    | Left x -> f x
    | Right y -> g y
end

open Either

let contains s1 s2 =
  try
    let len = String.length s2 in
    for i = 0 to String.length s1 - len do
      if String.sub s1 i len = s2 then raise Exit
    done;
    false
  with Exit -> true

let haspink = fun s ->
  contains s "pink"

let hasrose = fun s ->
  contains s "rose"

let isvalid s =
  haspink s || hasrose s

let solve cs =
  List.map String.lowercase_ascii cs
  |> List.filter isvalid
  |> List.length
  |> fun n ->
  if n=0 then Right "I must watch Star Wars with my daughter"
  else Left n

let iden x = x

let () =
  let t = read_int () in
  let rec readt t =
    if t=0 then []
    else read_line () :: readt (t-1) in
  readt t |> solve |> either string_of_int iden |> print_endline
